src/index.js

```diff
@@ -0,0 +1,196 @@
+
+    let factory;
+    let element;
+    if (t.react.isCompatTag(jsxElementName.name)) {
+      factory = "element";
+      element = jsxToString(jsxElementName);
+    } else {
+      factory = "component";
+      element = jsxToJS(jsxElementName);
+    }
+
+    return t.callExpression(
+      t.memberExpression(t.identifier("ESXToken"), t.identifier(factory)),
+      [
+        element,
+        transformAttributesList(path.get("openingElement")),
+        ...path.get("children").map(transformChild).filter(Boolean),
+      ]
+    );
+  }
+
+  function transformFragment(path) {
+    return t.callExpression(
+      t.memberExpression(t.identifier("ESXToken"), t.identifier("factory")),
+      path.get("children").map(transformChild).filter(Boolean)
+    );
+  }
+
+  /**
+   * @param {import("@babel/types").JSXIdentifier | import("@babel/types").JSXMemberExpression} node
+   * @returns {import("@babel/types").Identifier | import("@babel/types").MemberExpression}
+   */
+  function jsxToJS(node) {
+    if (t.isJSXMemberExpression(node)) {
+      return t.inherits(
+        t.memberExpression(jsxToJS(node.object), jsxToJS(node.property)),
+        node
+      );
+    }
+    return t.inherits(t.identifier(node.name), node);
+  }
+
+  /**
+   * @param {import("@babel/types").JSXIdentifier} node
+   * @returns {import("@babel/types").StringLiteral}
+   */
+  function jsxToString(node) {
+    return t.inherits(t.stringLiteral(node.name), node);
+  }
+
+  function transformAttributesList(path) {
+    /** @type {import("@babel/types").JSXOpeningElement} */
+    const node = path.node;
+
+    if (node.attributes.length === 0) return t.nullLiteral();
+
+    let hasSpread, hasDynamic, hasStatic;
+    for (const attr of node.attributes) {
+      if (t.isJSXSpreadAttribute(attr)) {
+        hasSpread = true;
+        break;
+      } else if (
+        t.isJSXAttribute(attr) &&
+        t.isJSXExpressionContainer(attr.value)
+      ) {
+        hasDynamic = true;
+      } else {
+        hasStatic = true;
+      }
+    }
+    const type =
+      hasSpread || !hasStatic
+        ? "RUNTIME_TYPE"
+        : hasDynamic
+        ? "MIXED_TYPE"
+        : "STATIC_TYPE";
+
+    return template.expression.ast`
+      ESXToken.create(
+        ESXToken.${t.identifier(type)},
+        ${t.arrayExpression(path.get("attributes").map(transformAttribute))}
+      )
+    `;
+  }
+
+  function transformAttribute(path) {
+    /** @type {import("@babel/types").JSXAttribute | import("@babel/types").JSXSpreadAttribute} */
+    const node = path.node;
+
+    let type, name, value;
+    if (t.isJSXSpreadAttribute(node)) {
+      type = "RUNTIME_TYPE";
+      name = t.stringLiteral("TODO: spread");
+      value = node.argument;
+    } else if (t.isJSXExpressionContainer(node.value)) {
+      type = "RUNTIME_TYPE";
+      name = jsxToString(node.name);
+      value = node.value.expression;
+    } else if (t.isJSXElement(node.value) || t.isJSXFragment(node.value)) {
+      throw path
```

Comment:
Do you want to support `<div x=<b/>`? What should the output look like?

Reply:
Nope, unless is a common use case in JSX but I don’t see any value in supporting that